### ISSUE TITLE
`x is None` type guard incorrectly narrows `Self` in the negative case

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#6361]
The `x is None` type guard incorrectly narrows the `Self` type (or any bound TypeVar) in the negative (else) case.

This comes from a mypy bug report: https://github.com/python/mypy/issues/16418

Here's a copy of the code sample that exhibits the false positive:

```python
from typing import Self


class A:
    def __init__(self, parent: Self | None) -> None:
        self.parent = parent

    def get_depth(self) -> int:
        current: Self | None = self
        count = 0
        while current is not None:
            count += 1
            current = current.parent
        return count - 1
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.335, which I just published. It will also be included in a future release of pylance.
